netket/operator/_graph_operator.py

```diff
@@ -5,87 +5,86 @@
-            operator. If None, the graph is deduced from the hilbert object.
-     site_ops: A list of operators in matrix form that act
-            on the nodes of the graph.
-            The default is an empty list. Note that if no site_ops are
-            specified, the user must give a list of bond operators.
-     bond_ops: A list of operators that act on the edges of the graph.
-         The default is None. Note that if no bond_ops are
-         specified, the user must give a list of site operators.
-     bond_ops_colors: A list of edge colors, specifying the color each
-         bond operator acts on. The defualt is an empty list.
-
-    Examples:
-     Constructs a ``BosGraphOperator`` operator for a 2D system.
-
-     >>> import netket as nk
-     >>> sigmax = [[0, 1], [1, 0]]
-     >>> mszsz = [[1, 0, 0, 0], [0, -1, 0, 0], [0, 0, -1, 0], [0, 0, 0, 1]]
-     >>> edges = [[0, 1], [1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8],
-     ... [8, 9], [9, 10], [10, 11], [11, 12], [12, 13], [13, 14], [14, 15],
-     ... [15, 16], [16, 17], [17, 18], [18, 19], [19, 0]]
-     >>> g = nk.graph.CustomGraph(edges=edges)
-     >>> hi = nk.hilbert.CustomHilbert(local_states=[-1, 1], graph=g)
-     >>> op = nk.operator.GraphOperator(
-     ... hi, site_ops=[sigmax], bond_ops=[mszsz])
-     >>> print(op.hilbert.size)
-     20
-    """
-
-    assert (
-        graph.n_nodes == hilbert.size
-    ), "The size of the graph must match the hilbert space"
-
-    size = graph.n_nodes
-
-    # Create the local operator as the sum of all site and bond operators
-
-    # Ensure that at least one of SiteOps and BondOps was initialized
-    if len(bond_ops) == 0 and len(site_ops) == 0:
-        raise InvalidInputError("Must input at least site_ops or bond_ops.")
-
-    this_operator = LocalOperator(hilbert)
-
-    # Site operators
-    if len(site_ops) > 0:
-        for i in range(size):
-            for j, site_op in enumerate(site_ops):
-                this_operator += LocalOperator(hilbert, site_op, [i])
-
-    # Bond operators
-    if len(bond_ops_colors) > 0:
-        if len(bond_ops) != len(bond_ops_colors):
-            raise InvalidInputError(
-                """The GraphHamiltonian definition is inconsistent.
-                The sizes of bond_ops and bond_ops_colors do not match."""
-            )
-
-        if len(bond_ops) > 0:
-            #  Use edge_colors to populate operators
-            for (u, v, color) in graph.edges(color=True):
-                edge = u, v
-                for c, bond_color in enumerate(bond_ops_colors):
-                    if bond_color == color:
-                        this_operator += LocalOperator(hilbert, bond_ops[c], edge)
-    else:
-        assert len(bond_ops) == 1
-
-        for edge in graph.edges():
-            this_operator += LocalOperator(hilbert, bond_ops[0], edge)
-
-    return this_operator
+class GraphOperator(LocalOperator):
```

Comment:
Is this now a class only to have a custom `__repr__` method?

Reply:
I strongly dislike that, but I am not so experienced in Python to know if this is justified or not..
Can't we override `__repr__` in a LocalOperator to report back that it is a graph operator?

Something along the lines of 

```python
def GraphOperator(hilbert, graph....):
  this_opereator = LocalOperator(hilbert)
  # do stuff...

  def custom_graph_repr(self):
      return "graphOperator(....)"

   this_operator.__repr__ = custom_graph_repr_
   return this_operator
```

I think there should be a way to do this in python.
I don't know if fit's a good idea...

Reply:
While it is technically possible to "hot-patch" a different methods to objects, this actually does not work with `repr` (or, I think, similar functions such as `len`). This is because
`repr(obj)` is turned into `Class.__repr__(obj)` if `obj` is of class `Class`, instead of `obj.__repr__()`, so the patched version is ignored unless it is defined on the class (meaning one has to create a subclass).

Also, at least for `GraphOperator`, it think it can make sense for it to be a subclass for other reasons, since there might be cases were it makes sense to add or override methods that make use of the graph structure. For example, right now we store a local operator for each edge
```python
>>> hi = Spin(s=1/2, N=100)
>>> ham = Heisenberg(hi, Chain(100))
>>> len(ham._operators)
100
```
even though they are all equal and would only need to be stored once. Making these kinds of optimizations in the future would at least be easier if `GraphOperator` is a class of its own.